lib/normalizer.c

```diff
@@ -909,15 +902,53 @@ grn_nfkc_normalize_unify_alphabet_diacritical_mark(
        * U+1EB3 LATIN SMALL LETTER A WITH BREVE AND HOOK ABOVE
        * U+1EB5 LATIN SMALL LETTER A WITH BREVE AND TILDE
        * U+1EB7 LATIN SMALL LETTER A WITH BREVE AND DOT BELOW
-       * Uppercase counterparts(U+1EA2, U+1EA4, U+1EA6, U+1EA8, U+1EAA,
+       * Uppercase counterparts (U+1EA2, U+1EA4, U+1EA6, U+1EA8, U+1EAA,
        * U+1EAC, U+1EAE, U+1EB0, U+1EB2, U+1EB4, and U+1EB6) are covered by the
        * following condition but they are never appeared here. Because NFKC
        * normalization converts them to their lowercase equivalents.
        */
       (utf8_char[0] == 0xE1 && utf8_char[1] == 0xBA &&
        (0xA1 <= utf8_char[2] && utf8_char[2] <= 0xB7))) {
+    return true;
+  }
+  return false;
+}
+
+grn_inline static bool
+grn_nfkc_normalize_unify_diacritical_mark_is_b(const unsigned char *utf8_char)
+{
+  /*
+   * Latin Extended Additional
+   * U+1E03 LATIN SMALL LETTER B WITH DOT ABOVE
+   * U+1E05 LATIN SMALL LETTER B WITH DOT BELOW
+   * U+1E07 LATIN SMALL LETTER B WITH LINE BELOW
+   * Uppercase counterparts (U+1E04, U+1E06) are covered by the
+   * following condition but they are never appeared here. Because NFKC
+   * normalization converts them to their lowercase equivalents.
+   */
+  if (utf8_char[0] == 0xE1 && utf8_char[1] == 0xB8 &&
+      (0x83 <= utf8_char[2] && utf8_char[2] <= 0x87)) {
+    return true;
+  }
+  return false;
+}
+
+/*
+ * This function assumes that the input utf8_char is a valid UTF-8 character.
+ * It is the caller's responsibility to ensure that utf8_char is valid UTF-8
+ * before calling this function. This is a precondition in terms of contract
+ * programming.
+ */
+grn_inline static const unsigned char *
+grn_nfkc_normalize_unify_alphabet_diacritical_mark(
+  const unsigned char *utf8_char, unsigned char *unified)
+{
+  if (grn_nfkc_normalize_unify_diacritical_mark_is_a(utf8_char)) {
     *unified = 'a';
     return unified;
+  } else if (grn_nfkc_normalize_unify_diacritical_mark_is_b(utf8_char)) {
+    *unified = 'b';
+    return unified;
   }
 
   return utf8_char;
```

Comment:
こうするなら`else`を作ってからなずどれかのブランチで処理されるっていうコードの方がすべてのパターンを処理したいんです！という意図が伝わるかな。

今のスタイルだとearly returnだからマッチしたときが特別扱いなんだけど、今回は全部同じ扱いじゃないかな。
参考： https://www.clear-code.com/blog/2012/3/28/best-practice-how-to-use-if-and-return.html

```suggestion
  } else {
    return utf8_char;
  }
```

Reply:
fix: [053188d](https://github.com/groonga/groonga/pull/1891/commits/053188dc2cb63f1277b8b40d85c347491f55a2c5) Diacrtical mark を取り除くという目的で同列の作業なため、`if else`を利用してすべて同じ扱いに変更しました。